File: tests/test_transform.py
"""Tests for src/transform.py"""
import json
from pathlib import Path

import pytest

from src.transform import (
    transform_bootstrap,
    transform_element_summary,
    transform_event_live,
    transform_fixtures,
)

FIXTURES = Path(__file__).parent / "fixtures"


def load(filename: str) -> dict | list:
    return json.loads((FIXTURES / filename).read_text())


class TestTransformBootstrap:
    def test_returns_correct_counts(self):
        data = load("bootstrap_static.json")
        teams, gameweeks, players = transform_bootstrap(data)
        assert len(teams) == 1
        assert len(gameweeks) == 2
        assert len(players) == 1

    def test_team_fields(self):
        data = load("bootstrap_static.json")
        teams, _, _ = transform_bootstrap(data)
        t = teams[0]
        assert t.fpl_id == 1
        assert t.name == "Arsenal"
        assert t.short_name == "ARS"
        assert t.strength == 4

    def test_gameweek_flags(self):
        data = load("bootstrap_static.json")
        _, gameweeks, _ = transform_bootstrap(data)
        current = next(gw for gw in gameweeks if gw.is_current)
        assert current.fpl_id == 25
        finished = next(gw for gw in gameweeks if gw.is_finished)
        assert finished.fpl_id == 1

    def test_player_fields(self):
        data = load("bootstrap_static.json")
        _, _, players = transform_bootstrap(data)
        p = players[0]
        assert p.fpl_id == 318
        assert p.web_name == "Salah"
        assert p.now_cost == 130
        assert p.total_points == 185
        assert p.element_type == 3

    def test_player_numeric_string_fields(self):
        data = load("bootstrap_static.json")
        _, _, players = transform_bootstrap(data)
        p = players[0]
        # These stay as strings (FPL returns them as strings)
        assert p.form == "12.8"
        assert p.selected_by_percent == "45.2"
        assert p.expected_goals == "15.34"

    def test_missing_optional_fields(self):
        """Bootstrap with minimal required fields should not raise."""
        data = {
            "teams": [{"id": 1, "name": "Test FC", "short_name": "TST"}],
            "events": [],
            "elements": [],
        }
        teams, gameweeks, players = transform_bootstrap(data)
        assert len(teams) == 1
        assert teams[0].strength is None

    def test_malformed_player_is_skipped(self):
        """A player missing required 'id' field should be skipped, not raise."""
        data = {
            "teams": [],
            "events": [],
            "elements": [
                {"web_name": "NoId", "team": 1, "element_type": 3}  # missing id
            ],
        }
        _, _, players = transform_bootstrap(data)
        assert len(players) == 0


class TestTransformElementSummary:
    def test_history_row_count(self):
        data = load("element_summary_318.json")
        history, history_past = transform_element_summary(318, data)
        assert len(history) == 1
        assert len(history_past) == 1

    def test_history_fields(self):
        data = load("element_summary_318.json")
        history, _ = transform_element_summary(318, data)
        h = history[0]
        assert h.player_fpl_id == 318
        assert h.gameweek_fpl_id == 24
        assert h.total_points == 14
        assert h.goals_scored == 2
        assert h.was_home == 1

    def test_history_past_fields(self):
        data = load("element_summary_318.json")
        _, history_past = transform_element_summary(318, data)
        hp = history_past[0]
        assert hp.player_fpl_id == 318
        assert hp.season_name == "2023/24"
        assert hp.total_points == 229


class TestTransformFixtures:
    def test_fixture_count(self):
        data = load("fixtures.json")
        fixtures = transform_fixtures(data)
        assert len(fixtures) == 1

    def test_fixture_fields(self):
        data = load("fixtures.json")
        fixtures = transform_fixtures(data)
        f = fixtures[0]
        assert f.fpl_id == 1
        assert f.gameweek_fpl_id == 1
        assert f.team_h_fpl_id == 14
        assert f.team_a_fpl_id == 3
        assert f.finished == 1
        assert f.team_h_score == 1

    def test_fixture_stats_serialised_as_json(self):
        data = load("fixtures.json")
        fixtures = transform_fixtures(data)
        import json
        stats = json.loads(fixtures[0].stats)
        assert stats[0]["identifier"] == "goals_scored"


class TestTransformEventLive:
    def test_live_row_count(self):
        data = load("event_25_live.json")
        rows = transform_event_live(25, data)
        assert len(rows) == 1

    def test_live_fields(self):
        data = load("event_25_live.json")
        rows = transform_event_live(25, data)
        r = rows[0]
        assert r.player_fpl_id == 318
        assert r.gameweek_fpl_id == 25
        assert r.total_points == 12
        assert r.in_dreamteam == 1
        assert r.minutes == 90

    def test_explain_serialised_as_json(self):
        data = load("event_25_live.json")
        rows = transform_event_live(25, data)
        import json
        explain = json.loads(rows[0].explain)
        assert explain[0]["fixture"] == 290

File: tests/test_transform.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { beforeEach, describe, expect, it, vi } from "vitest";

const { debugSpy, warnSpy } = vi.hoisted(() => ({
  debugSpy: vi.fn(),
  warnSpy: vi.fn(),
}));

vi.mock("../src/logger.ts", () => ({
  getLogger: () => ({
    debug: debugSpy,
    warn: warnSpy,
  }),
}));

import {
  transform_bootstrap,
  transform_element_summary,
  transform_event_live,
  transform_fixtures,
} from "../src/transform.ts";

const FIXTURES_DIR = join(dirname(fileURLToPath(import.meta.url)), "fixtures");

function loadFixture<T>(filename: string): T {
  return JSON.parse(readFileSync(join(FIXTURES_DIR, filename), "utf8")) as T;
}

beforeEach(() => {
  debugSpy.mockClear();
  warnSpy.mockClear();
});

describe("src/transform.ts", () => {
  it("transforms bootstrap fixtures into team, gameweek, and player rows", () => {
    const data = loadFixture<Record<string, unknown>>("bootstrap_static.json");

    const [teams, gameweeks, players] = transform_bootstrap(data);

    expect(teams).toHaveLength(1);
    expect(gameweeks).toHaveLength(2);
    expect(players).toHaveLength(1);

    expect(teams[0]).toMatchObject({
      fpl_id: 1,
      name: "Arsenal",
      short_name: "ARS",
      strength: 4,
    });
    expect(
      gameweeks.find((gameweek) => gameweek.is_current === 1),
    ).toMatchObject({
      fpl_id: 25,
    });
    expect(players[0]).toMatchObject({
      fpl_id: 318,
      web_name: "Salah",
      now_cost: 130,
      total_points: 185,
      element_type: 3,
      form: "12.8",
      selected_by_percent: "45.2",
      expected_goals: 15.34,
    });
    expect(debugSpy).toHaveBeenCalledWith(
      "Transformed bootstrap: %d teams, %d gameweeks, %d players",
      1,
      2,
      1,
    );
  });

  it("treats missing bootstrap collections as empty and skips malformed players with warnings", () => {
    const data = loadFixture<Record<string, unknown>>("bootstrap_static.json");
    const malformedBootstrap = {
      ...data,
      teams: undefined,
      events: undefined,
      elements: [
        ...(data.elements as unknown[]),
        { web_name: "NoId", team: 1, element_type: 3 },
      ],
    };

    const [teams, gameweeks, players] = transform_bootstrap(malformedBootstrap);

    expect(teams).toEqual([]);
    expect(gameweeks).toEqual([]);
    expect(players).toHaveLength(1);
    expect(players[0].fpl_id).toBe(318);
    expect(warnSpy).toHaveBeenCalledWith(
      "Skipping malformed player %s: %s",
      undefined,
      expect.any(TypeError),
    );
  });

  it("transforms element-summary fixtures into current and past history rows", () => {
    const data = loadFixture<Record<string, unknown>>(
      "element_summary_318.json",
    );

    const [history, historyPast] = transform_element_summary(318, data);

    expect(history).toHaveLength(1);
    expect(historyPast).toHaveLength(1);
    expect(history[0]).toMatchObject({
      player_fpl_id: 318,
      gameweek_fpl_id: 24,
      total_points: 14,
      goals_scored: 2,
      was_home: 1,
    });
    expect(historyPast[0]).toMatchObject({
      player_fpl_id: 318,
      season_name: "2023/24",
      total_points: 229,
    });
  });

  it("treats missing element-summary arrays as empty collections", () => {
    const [history, historyPast] = transform_element_summary(318, {});

    expect(history).toEqual([]);
    expect(historyPast).toEqual([]);
    expect(warnSpy).not.toHaveBeenCalled();
  });

  it("transforms fixtures fixtures into stable rows and preserves serialized stats blobs", () => {
    const data = loadFixture<unknown>("fixtures.json");

    const fixtures = transform_fixtures(data);

    expect(fixtures).toHaveLength(1);
    expect(fixtures[0]).toMatchObject({
      fpl_id: 1,
      gameweek_fpl_id: 1,
      team_h_fpl_id: 14,
      team_a_fpl_id: 3,
      finished: 1,
      team_h_score: 1,
    });
    expect(JSON.parse(fixtures[0].stats as string)).toEqual([
      {
        identifier: "goals_scored",
        h: [{ value: 1, element: 200 }],
        a: [{ value: 1, element: 150 }],
      },
    ]);
    expect(debugSpy).toHaveBeenCalledWith("Transformed %d fixtures", 1);
  });

  it("returns an empty fixture list when the payload is not an array", () => {
    expect(transform_fixtures({ fixtures: [] })).toEqual([]);
    expect(warnSpy).not.toHaveBeenCalled();
  });

  it("transforms event-live fixtures into live stat rows and preserves nested JSON blobs", () => {
    const data = loadFixture<Record<string, unknown>>("event_25_live.json");

    const rows = transform_event_live(25, data);

    expect(rows).toHaveLength(1);
    expect(rows[0]).toMatchObject({
      player_fpl_id: 318,
      gameweek_fpl_id: 25,
      total_points: 12,
      in_dreamteam: 1,
      minutes: 90,
      influence: "60.2",
      expected_goals_conceded: "0.74",
    });
    expect(JSON.parse(rows[0].explain as string)).toEqual([
      {
        fixture: 290,
        stats: [
          { identifier: "minutes", points: 2, value: 90 },
          { identifier: "goals_scored", points: 5, value: 1 },
          { identifier: "assists", points: 3, value: 1 },
          { identifier: "bonus", points: 3, value: 3 },
        ],
      },
    ]);
    expect(debugSpy).toHaveBeenCalledWith(
      "Transformed %d live stats rows for GW%d",
      1,
      25,
    );
  });

  it("treats missing event-live elements as an empty collection", () => {
    expect(transform_event_live(25, {})).toEqual([]);
    expect(warnSpy).not.toHaveBeenCalled();
  });
});
